# Homework 1
*Work done by Hector A. Rodriguez and Siddhanth Lalgowdar*

## Question 1
In class we discussed how to learn a model of the form 
y = w_0 + w_1(x) when w_0 = A is known. 
The model was then used to predict the numeric value f(x) of an arbitrary input scalar x, where f is the unknown "ground truth".

(a) Expand the model to the case where both w_0 and w_1 are unknown and write a matrix equation for the optimaly parameter vector **w** that minimizes the cost.

(b) Using NumPy, write code to find the optimal **w** and test it on the small data set used in class. Does it give a better solution? Submit a Jupyter Notebook showing your work.

### Answers 
**(a) -**
In Advanced Linear Algebra, there is the Projection Theorem, which helps calculate coefficients of matrices. 

For every **w**, we want to make sure that ||**y** - **y_hat**|| ≤ ||**y** - **w**||

Since we know that **w** = A**x**, we can try to solve for ||**y** - A**x**||

With this, we can find the **x_hat** = (A^t * A)^(-1) * A^t **y**.

And **y_hat** = A * (A^t * A)^(-1) * A^t **y**.

In the case of our question, **x_hat** would be our coefficient vector = \[w_0  w_1\], A would be our data \[ **x_1**  ...  **x_N** \], and **y_hat** would be our usual **y_hat**

So the final equation looks something like this:

**w** = (**x**^t * **x**)^(-1) **x**^t **y**

**(b) -** Yes, I would say that this method gives a better solution, with slightly better accruacy 

In [14]:
import numpy as np
import pandas as pd

d = {'y': [12, 12, 13, 18], 'x': [3, 4, 5, 6]}
data = pd.DataFrame(data = d)

## Remove and save the y
y = data["y"]
del data["y"]

## Add column of 1 for w_0
data.insert(0, "bias", 1)

linReg = data.to_numpy()

## Use Projection Theorem for coeff
## coeff[0] -> w_0 (coefficient bias)
## coeff[1] -> w_1 (coefficient for x)
coeff = np.linalg.inv(linReg.T@linReg)@linReg.T@y
coeff = coeff.round(2)
print(f'w_0: {coeff[0]}\nw_1: {coeff[1]}')

## Calculate y_hat
predicted = linReg@coeff
pd.DataFrame({"Target / y": y, "Predicted / y_hat": predicted})

w_0: 5.2
w_1: 1.9


Target / y  Predicted / y_hat
0          12               10.9
1          12               12.8
2          13               14.7
3          18               16.6

## Question 2
In class we discussed a greedy divide-and-conquer learning algorithm for binary classifiaction. The algorithm produces a model *g* that predicts the class of input **x**. In the following questions, assume all features are binary.

(a) What is a good representation for the model *g*?

(b) Under the empirical risk management (ERM), what is the learned model *g* for the data set presented in class? Provide the simplest possible ERM model. 

(c) What is the training error of *g*?

(d) When is the ERM algorithm guaranteed to return a training error equal to zero (again under ERM) and when not? Is it always a good idea to minimize the training error? Explain.

(e) (**grad**) We are not grad students, so we are not doing this.

### Answers
**(a) -**
A good representation for the model *g* is a binary tree, in which each node would be a feature, and the branches depict the binary feature options. 

**(b) -**
Under ERM, the simplest possible ERM model is a binary tree with the feature **Systems** as the head. 

**(c) -**
The training error of *g* chosen above is 2/12 students who say no to **Systems** are misclassified and 0/8 students who say yes to **Systems** are misclassified, for a total error rate of *16.6%* error rate.

**(d) -**
The ERM algorithm is guaranteed to return a training error equal to zero when there exists a model within the hypothesis space that can overfit the data given. It it not a good idea to minimize the training error in this case, because then you have an overfit model that cannot generalize to other data that it has not seen / been trained on. 

**(e) -**
*Does Not Exist*

## Question 3

a.

The probability of a point from sample space falling into the gap is E.
The probability of a point falling outside the gap is 1 - E.
The probability of N points falling outside the gap is (1 - E)^N.

We know that (1 - E) = 1 + E + E^2/2! + E^3/3! + ... = e^-E

Therefore, (1 - E)^N  = e^-EN = delta

-EN ln e = ln delta

N = -ln delta/E

b.

c.

1 - delta = 99% => delta = 0.01
E = 5% = 0.05

N = -ln (0.01) / 0.05 = 92

## Question 4

Scikit-learn uses this formula to calculate idf:

idf = log(n_samples / df(w)) + 1.0

In [34]:
import sklearn
import numpy as np
import math

In [ ]:
sentences = ['The fat cat sat in the hat',
             'The cat destroyed the hat',
             'The hat is too big',
             'The fat cat likes the big cat',
            'The big dog likes fat people']



In [18]:
vocabulary = ['big', 'cat', 'destroyed', 'dog', 'fat', 'hat', 'in', 'likes', 'people', 'sat', 'too']

sentences = ['The fat cat sat in the hat',
             'The cat destroyed the hat',
             'The hat is too big',
             'The fat cat likes the big cat',
            'The big dog likes fat people']

vocab_mat = []

## bag-of-words encoding

for s in sentences:
    vocab_li = [0]*11
    words = s.split(' ')
    for w in words:
        if w in vocabulary:
            vocab_li[vocabulary.index(w)] += 1
    vocab_mat.append(vocab_li)

vocab_mat

[[0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0],
 [0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0],
 [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
 [1, 2, 0, 0, 1, 0, 0, 1, 0, 0, 0],
 [1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0]]

In [39]:
idf_mat = []
for v in vocab_mat:
    for i in v:
        if i != 0:
            print(math.log((len(vocab_mat))/i))


1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
0.9162907318741551
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
1.6094379124341003
